test/topology_experimental_raft/test_alternator.py

```diff
@@ -0,0 +1,203 @@
+#
+# SPDX-License-Identifier: AGPL-3.0-or-later
+#
+
+# Multi-node tests for Alternator.
+#
+# Please note that most tests for Alternator are single-node tests and can
+# be found in the test/alternator directory. Most functional testing of the
+# many different syntax features that Alternator provides don't need more
+# than a single node to be tested, and should be able to run also on DynamoDB
+# - not just on Alternator, which the test/alternator framework allows to do.
+# So only the minority of tests that do need a bigger cluster should be here.
+
+import pytest
+import asyncio
+import logging
+import time
+import boto3
+import botocore
+import requests
+import json
+
+from test.pylib.manager_client import ManagerClient
+
+logger = logging.getLogger(__name__)
+
+# Convenience function to open a connection to Alternator usable by the
+# AWS SDK.
+alternator_config = {
+    'alternator_port': 8000,
+    'alternator_write_isolation': 'only_rmw_uses_lwt',
+    'alternator_ttl_period_in_seconds': '0.5',
+}
+def get_alternator(ip):
+    url = f"http://{ip}:{alternator_config['alternator_port']}"
+    return boto3.resource('dynamodb', endpoint_url=url,
+        region_name='us-east-1',
+        aws_access_key_id='alternator',
+        aws_secret_access_key='secret_pass',
+        config=botocore.client.Config(
+            retries={"max_attempts": 0},
+            read_timeout=300)
+    )
+
+# Alternator convenience function for fetching the entire result set of a
+# query into an array of items.
+def full_query(table, ConsistentRead=True, **kwargs):
+    response = table.query(ConsistentRead=ConsistentRead, **kwargs)
+    items = response['Items']
+    while 'LastEvaluatedKey' in response:
+        response = table.query(ExclusiveStartKey=response['LastEvaluatedKey'],
+            ConsistentRead=ConsistentRead, **kwargs)
+        items.extend(response['Items'])
+    return items
+
+# FIXME: boto3 is NOT async. So all tests that use it are not really async.
+# We could use the aioboto3 library to write a really asynchronous test, or
+# implement an async wrapper to the boto3 functions ourselves (e.g., run them
+# in a separate thread) ourselves.
+
+@pytest.fixture(scope="module")
+async def alternator3(manager_internal):
+    """A fixture with a 3-node Alternator cluster that can be shared between
+       multiple tests. These test should not modify the cluster's topology,
+       and should each use unique table names and/or unique keys to avoid
+       being confused by other tests.
+       Returns the manager object and 3 boto3 resource objects for making
+       DynamoDB API requests to each of the nodes in the Alternator cluster.
+    """
+    manager = manager_internal()
+    servers = await manager.servers_add(3, config=alternator_config)
+    yield [manager] + [get_alternator(server.ip_addr) for server in servers]
+    await manager.stop()
+
+test_table_prefix = 'alternator_Test_'
+def unique_table_name():
```

Comment:
really don't see why 2 python projects can't share code and we need to have a set of helpers for test/alternator and alternator in topology

Reply:
I'm sad that you (and others) still don't understand my plea to **not** share code between test/alternator and test/topology, and why it is a mistake to share code between pieces of the project which don't share the same ideals and same style.

Please spend some time looking at test/pylib and count the number of lines there, and tell me if it's something I want to see anywhere NEAR test/alternator with its (deliberately) minimal library of convenience functions.

Moreover, test/alternator/util.py is 240 lines. That's 3% (!!!) the size of test/pylib. Even if I duplicate all of it, it will be completely negligible compared to the existing test/pylib.

I think the `topology*/*` tests are in a bad state, perhaps not as bad as dtest but still bad, it encourages writing slow tests that needlessly start nodes again and again, writing hard-to-read tests with over-engineered programming practices that are't needed for short tests, and so on. There is also no excuse for why so many different directories exist. Until these things are sorted out, I don't want any of these mistakes to "infiltrate" test/cql-pytest (and its 1507 tests) or test/alternator (765 tests) and want them to evolve independently.
All `topology*` directories now have a total of 141 tests. It's still a baby.

Reply:
Yes, I don't agree with this duplication (long term it doesn't help, see how topology duplicated some functionality of dtest and now we have two piles of code instead of one).

But I can live with the duplication you're adding, as you calculated 3% is not much.


> I think the topology*/* tests are in a bad state, perhaps not as bad as dtest but still bad, it encourages writing slow tests that needlessly start nodes again and again

To some extent I agree, but the thing is to somehow make it better not start your own sandbox inside of them.

Reply:
> Yes, I don't agree with this duplication (long term it doesn't help, see how topology duplicated some functionality of dtest and now we have two piles of code instead of one).

That duplication (topology vs. dtest) was necessary for innovation. Without it, we would have JUST dtest. I'm happy that code was not reused.

> To some extent I agree, but the thing is to somehow make it better not start your own sandbox inside of them.

Perhaps, but it's too hard to change everything (there's a huge amount of code in test/pylib) so what's my other option - if I can't force everyone to improve their bad practices, I should adopt these bad practices too?